### ISSUE TITLE
common table expression in inner query

### ISSUE THREAD

**Issue by xiangjinwu** [risingwavelabs/risingwave#3174]
Some advanced issues that can be ignored in this first PR:
* The HashMap storing ctes should be part of the hierarchical context rather than global in binder:
```
with cte0 as (select 0)
select * from
    (
        with cte1 as (select 1)
        select from cte0, cte1  -- cte0 from outside can be used here
    ) as sub0, cte0;  -- cte1 from inside cannot be used here
```
* When a cte has the same name as a table, it take [precedence](https://www.postgresql.org/docs/current/sql-select.html#:~:text=the%20WITH%20query%20hides%20any%20real%20table%20of%20the%20same%20name%20for%20the%20purposes%20of%20the%20primary%20query.%20If%20necessary%2C%20you%20can%20refer%20to%20a%20real%20table%20of%20the%20same%20name%20by%20schema%2Dqualifying%20the%20table%27s%20name.), unless the name is qualified with schema name. (handled by #3155)

_Originally posted by @xiangjinwu in https://github.com/singularity-data/risingwave/issues/2741#issuecomment-1137019991_

**Comment by TennyZhuang** [comment]
So cte0 can use the outer columns so the query may be a correlated subquery?

**Comment by chenzl25** [comment]
I provide a case here.
```sql
with c as (values(1)) select * from (with c as (values(2)) select * from c) foo union all select * from c;
```

Expected result:
```
 column1
---------
       2
       1
(2 rows)
```

Actual result:
```
 column1
---------
       2
       2
(2 rows)
```

**Comment by chenzl25** [comment]
Fixed by #7212 .
